### PR TITLE
Meson changes

### PR DESCRIPTION
**Pull request by Apache-HB**
This pr rewrites most of the build process save for the `eir` directory. this pr relies on its accompanying [bootstrap changes](https://github.com/managarm/bootstrap-managarm/pull/97)
* Use `default_options` for basic config such as `cpp_std`
* Enable `warning_level=3`
* Offload configuration of subprojects to mesons layered cross files
* Remove all `add_global_parameters` and `add_project_parameters` as cross files can also handle these
* Reuse generators rather than creating new ones for each file
* Use `declare_dependency` and its fields `sources` and `dependencies` more liberally to reduce recreation of proto/bragi files and to reduce the amount of dependencies that need to be explicitly declared per target
* Remove duplicate calls to `find_program`, `get_option`, `subproject`, and `dependency` as was reasonable
* Formatting pass

These changes should make it easier to enable lto and unity builds as well as reducing the build/configure time

### CODE REVIEW

**Comment by Apache-HB** [comment]
the CI is failing because these changes depend on changes in bootstrap and the CI pipeline is using master rather than my branch

File: meson.build [hunk 1]
```diff
@@ -72,79 +102,75 @@ endif
...
+# when building these tools make sure they stay below everything else
+# as they depend on parts above
 if get_option('build_tools')
-	frigg = subproject('frigg')
-	frigg_dep = frigg.get_variable('frigg_dep')
-
-	subdir('tools/ostrace')
-	subdir('tools/bakesvr')
-	subdir('tools/pb2frigg')
+	foreach tool : [ 'ostrace', 'basesvr', 'pb2frigg' ]
```
Comment (Geertiebear) [inline meson.build:173]: ``basesvr`` seems like a typo here no?

**Comment by avdgrinten** [comment]
Regarding
> Use declare_dependency and its fields sources and dependencies more liberally to reduce recreation of proto/bragi files and to reduce the amount of dependencies that need to be explicitly declared per target
IMHO it would be great if the protocols libraries (for example protocols/mbus) would install the generated header files for this purpose. However, I couldn't make that make with Meson and resorted to duplicated generation. If it can be fixed, that'd be great.

**Comment by Apache-HB** [comment]
I think that should be all the big lists of sources and headers sorted

File: protocols/clock/meson.build [hunk 2]
```diff
@@ -1,28 +1,13 @@
...
-clock_proto_inc = include_directories('include/')
-#libclock_protocol = shared_library('clock_protocol', [clock_pb],
-#	dependencies: [lib_helix_dep, proto_lite_dep],
-#	include_directories: hw_proto_inc,
-#	install: true)
-
-clock_protocol_dep = declare_dependency(
-#	link_with: libhw_protocol,
-	include_directories: clock_proto_inc)
+clock_dep = declare_dependency(
```
Comment (avdgrinten) [inline protocols/clock/meson.build:6]: It is probably a good idea to keep the `_protocol_` (not only here, but for all protocols). For example, there is the mbus server and the mbus protocol. However, we could shorten it to `_proto_` or even `_prt_` or something similar.

**Review by avdgrinten — COMMENTED** [review]
Thank you for the PR; here is now a proper review. First of all, the PR cleans up a lot of files, most notably the generators. That is great! My main criticism, however, is about the configuration that has been moved to cross files. See the individual comments below for more details.

It would be great to merge this PR quickly -- it has already accumulated a few merge conflicts and the number of conflicts will only increase over time.

File: meson.build [hunk 3]
```diff
@@ -1,69 +1,99 @@
-project('managarm')
+project('managarm', ['c', 'cpp'],
+	license : 'MIT',
+	meson_version : '>=0.57.0',
+	default_options : [
+		'cpp_std=c++20',
+		'c_std=c11',
+		'warning_level=3'
```
[the remark this answers is not in the document]
Comment (Apache-HB) [reply meson.build:7]: warning_level 3 translates to `-Wall -Winvalid-pch -Wextra -Wpedantic` on gcc and clang

File: meson.build [hunk 4]
```diff
@@ -1,69 +1,99 @@
...
 
-	add_global_arguments(
-		[
-			'-ffreestanding',
-		],
-		language: ['c', 'cpp'])
+cxxshim = subproject('cxxshim')
+coroutines = cxxshim.get_variable('clang_coroutine_dep')
 
-	add_global_arguments(['-fno-exceptions', '-fno-rtti'], language: 'cpp')
```
[the remark this answers is not in the document]
Comment (Apache-HB) [reply meson.build:48]: when building I found that to not be the case as this single file is responsible for building both `managarm-kernel` and `managarm-system`. `managarm-system` requires both exceptions and rtti to be enabled due to protobuf using them. In the related [bootstrap-managarm pull](https://github.com/managarm/bootstrap-managarm/pull/97/files#diff-d125d5e35080706355dbe4a16fea83eae2ac5e58fc7eb6382d6902d628b92463) I disable them in the cross file thats always used when building `managarm-kernel`.

[the remark this answers is not in the document]
Comment (avdgrinten) [reply meson.build:48]: That's true (exceptions and RTTI are needed in managarm-system), but I would just add them to `args` in `eir` and `thor`.

File: meson.build (hunk 3 above)
[the remark this answers is not in the document]
Comment (avdgrinten) [reply meson.build:7]: Doesn't that mean that we should change "c++20" to "gnu+20" (three lines above) to avoid lots of `-Wpedantic` errors?

**Review by avdgrinten — APPROVED** [review]
LGTM but there is one comment remaining on the bootstrap PR.

**Comment by avdgrinten** [comment]
Thank you very much for the PR. I will squash all comments since it is not clear whether intermediate stage compile and there is a lot of churn.

**Pull request merged**